curvefs/test/metaserver/metaserver_s3_adaptor_test.cpp

```diff
@@ -0,0 +1,242 @@
+    resp_version.set_statuscode(::curvefs::metaserver::MetaStatusCode::OK);
+    EXPECT_CALL(mockMetaServerService_, UpdateInodeS3Version(_, _, _, _))
+        .WillRepeatedly(
+            DoAll(SetArgPointee<2>(resp_version),
+                  Invoke(S3RpcService_Version<UpdateInodeS3VersionRequest,
+                                              UpdateInodeS3VersionResponse>)));
+    std::set<std::string> uploadObject;
+    std::function<int(std::string, const char*, uint64_t)> upload_object =
+        [&uploadObject](std::string name, const char* buf, uint64_t length) {
+            LOG(INFO) << "upload object, name:" << name;
+            uploadObject.insert(name);
+            std::string data(buf, length);
+            return data.length();
+        };
+    std::function<int(std::string, const char*, uint64_t)> append_object =
+        [&uploadObject](std::string name, const char* buf, uint64_t length) {
+            LOG(INFO) << "append object, name:" << name;
+            uploadObject.insert(name);
+            std::string data(buf, length);
+            return data.length();
+        };
+    EXPECT_CALL(mockClientS3Client_, Upload(_, _, _))
+        .WillRepeatedly(Invoke(upload_object));
+    EXPECT_CALL(mockClientS3Client_, Append(_, _, _))
+        .WillRepeatedly(Invoke(append_object));
+
+    // replace s3 delete
+    std::set<std::string> deleteObject;
+    std::function<int(std::string)> delete_object =
+        [&deleteObject](std::string name) {
+            LOG(INFO) << "delete object, name:" << name;
+            deleteObject.insert(name);
+            return 0;
+        };
+    EXPECT_CALL(mockMetaserverS3Client_, Delete(_))
+        .WillRepeatedly(Invoke(delete_object));
+
+    curvefs::metaserver::Inode inode;
+    InitInode(&inode);
+
+    /*
+    1. write 3MB+1 from 0; (write)
+    2. write 2MB+1 from 2MB+2; (overwrite）
+    3. write 1MB+1 from 4MB+3; (append)
+  */
+    char* buf;
+    uint64_t offset = 0;
+    uint64_t fileLen = 3 * 1024 * 1024 + 1;
+    buf = new char[fileLen];
+    memset(buf, 'a', fileLen);
+    clientS3ClientAdaptor_->Write(&inode, offset, fileLen - offset, buf);
+    inode.set_length(fileLen);
+    delete[] buf;
+
+    offset = 2 * 1024 * 1024 + 2;
+    fileLen = 2 * 1024 * 1024 + 2 + 2 * 1024 * 1024 + 1;
+    buf = new char[fileLen];
+    memset(buf, 'b', fileLen);
+    clientS3ClientAdaptor_->Write(&inode, offset, fileLen - offset, buf);
+    inode.set_length(fileLen);
+    delete[] buf;
+
+    offset = 4 * 1024 * 1024 + 3;
+    fileLen = 4 * 1024 * 1024 + 3 + 1 * 1024 * 1024 + 1;
+    buf = new char[fileLen];
+    memset(buf, 'c', fileLen);
+    clientS3ClientAdaptor_->Write(&inode, offset, fileLen - offset, buf);
+    inode.set_length(fileLen);
+    delete[] buf;
+
+    /*
+    1. write 3MB+1 from 0; (write)
+    2. write 2MB+1 from 2MB+2; (overwrite）
+    3. write 1MB+1 from 4MB+3; (append)
+    */
+
+    int ret = metaserverS3ClientAdaptor_->Delete(&inode);
```

Comment:
add an idempotence test here, like this :
first delete part of s3 chunks and failed, second delete will delete it success

Reply:
fixed